src/sinks/util/http.rs

```diff
@@ -696,30 +699,30 @@ impl ItemBatchSize<Event> for HttpJsonBatchSizer {
 }
 
 /// HTTP request builder for HTTP stream sinks using the generic `HttpService`
-pub trait HttpServiceRequestBuilder {
-    fn build(&self, body: Bytes) -> Request<Bytes>;
+pub trait HttpServiceRequestBuilder<T> {
+    fn build(&self, body: Bytes, metadata: T) -> Request<Bytes>;
 }
 
 /// Generic 'Service' implementation for HTTP stream sinks.
 #[derive(Clone)]
-pub struct HttpService<B> {
+pub struct HttpService<B, T: Send> {
     batch_service:
-        HttpBatchService<BoxFuture<'static, Result<Request<Bytes>, crate::Error>>, HttpRequest>,
+        HttpBatchService<BoxFuture<'static, Result<Request<Bytes>, crate::Error>>, HttpRequest<T>>,
     _phantom: PhantomData<B>,
 }
 
-impl<B> HttpService<B>
+impl<B, T: Send + 'static> HttpService<B, T>
 where
-    B: HttpServiceRequestBuilder + std::marker::Sync + std::marker::Send + 'static,
+    B: HttpServiceRequestBuilder<T> + std::marker::Sync + std::marker::Send + 'static,
 {
     pub fn new(http_client: HttpClient<Body>, http_request_builder: B) -> Self {
         let http_request_builder = Arc::new(http_request_builder);
 
-        let batch_service = HttpBatchService::new(http_client, move |req: HttpRequest| {
+        let batch_service = HttpBatchService::new(http_client, move |req: HttpRequest<T>| {
             let request_builder = Arc::clone(&http_request_builder);
 
             let fut: BoxFuture<'static, Result<http::Request<Bytes>, crate::Error>> =
-                Box::pin(async move { Ok(request_builder.build(req.payload)) });
+                Box::pin(async move { Ok(request_builder.build(req.payload, req.other_metadata)) });
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** since not all of the downstream users of this will need this extra metadata, I would suggest making the interface for the build function pass the `HttpRequest` directly, and the implementors can get what they need from it. It is a bit cleaner than passing around the `()` unit type.

 One option would be to create a getter,`get_additional_metadata() -> Option<T>` function to the impl block for `HttpRequest`. Then the implementors of the `build` trait function could call `req.get_additional_metadata()` 
 
 WDYT?‏

```suggestion
                Box::pin(async move { Ok(request_builder.build(req)) });
```

Reply:
I like it! 
- I've added getters for all fields
- renamed `other_metadata` -> `additional_metadata`
- wrapped generic typing for `additional_metadata` with an Option

Reply:
Thanks for making that change, it looks nicer now. I had a follow up comment on that so I'll close this one.